app/controllers/api/v1/project_preferences_controller.rb

```diff
@@ -2,10 +2,29 @@ class Api::V1::ProjectPreferencesController < Api::ApiController
   include PreferencesController
 
   require_authentication :all, scopes: [:project]
+  resource_actions :create, :update, :show, :index, :update_settings
+  extra_schema_actions :update_settings
   schema_type :json_schema
+  before_action :validate_records, only: [:update_settings]
+
+  def update_settings
+    upp = UserProjectPreference.where(project_id: params_for[:project_id],
+                                      user_id: params_for[:user_id]).first
+    upp.settings.merge! params_for[:settings]
+    upp.save!
+    render status: :ok, nothing: true
+  end
 
   private
 
+  def validate_records
+    User.find params_for[:user_id]
```

Comment:
I actually like that the authorization is in a `before_action`. Don't know why you need to find the user (you're not even storing it in a variable), but storing the project in `@project` and then doing `@project.user_project_preferences.find_by!(user_id: params[:user_id])` should work and keep the authorization here in a before filter.